### PR TITLE
[Minor] Enable continuation of training

### PR DESCRIPTION
**Pull request by weberpals**
## :microscope: Background

Enable to continue the training of a model as described in #828.

Also needed to address #1542 

## :crystal_ball: Key changes

Load model state from the last checkpoint and continue training with a new scheduler. The last learning rate is used as a starting point for the new scheduler.

ExponentialLR is used as new scheduler since continuing with OneCycleLR leads to issues. @ourownstory please let me know if you have an idea for a specific scheduler that make sense here. 

## :clipboard: Review Checklist
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, added docstrings and data types to function definitions.
- [x] I have added pytests to check whether my feature / fix works.

Please make sure to follow our best practices in the [Contributing guidelines](https://github.com/ourownstory/neural_prophet/blob/main/CONTRIBUTING.md).

### CODE REVIEW

**Comment by ourownstory** [comment]
@Constantin343 "The last learning rate is used as a starting point for the new scheduler."
Do you mean the LR found in the LR test, or the actual LR at the end of the cycle?

**Review by ourownstory — CHANGES_REQUESTED** [review]
Looks like you got a first version running - good work!

I left a few comments aiming to robustify the user settings and to encapsulate code in a similar manner to existing practices. 

Furthermore, I think we should avoid hard-coding the optimizer and scheduler for this functionality. I suggest you make them optionally user-settable (with reasonable defaults). And if so, then it may make sense to do the same for the regular optimizer and scheduler in the same manner.

**Comment by Constantin343** [comment]
> @Constantin343 "The last learning rate is used as a starting point for the new scheduler." Do you mean the LR found in the LR test, or the actual LR at the end of the cycle?

Yeah, the learning rate at the end of the cycle (actually from the last checkpoint, to be precise). The idea was to ensure a smooth transition

File: neuralprophet/forecaster.py [hunk 1]
```diff
@@ -1067,6 +1067,10 @@ def fit(
 
         if self.fitted is True and not continue_training:
             log.error("Model has already been fitted. Re-fitting may break or produce different results.")
+
+        if continue_training and self.metrics_logger.checkpoint_path is None:
+            log.error("Continued training requires checkpointing in model.")
```
[the remark this answers is not in the document]
Comment (Constantin343) [reply neuralprophet/forecaster.py:1072]: My thinking was that it makes sense to continue from the checkpoint, but probably it's not necessary. All the necessary parameters should still be available in the model itself. I will adapt it

[the remark this answers is not in the document]
Comment (Constantin343) [reply neuralprophet/forecaster.py:1072]: I did some testing and it seems like checkpoint is indeed necessary to correctly continue training with the pytroch-lighting trainer. I can get it to run without checkpointing but fitting again always leads to a complete restart of the training.
 
Maybe there is some workaround, but I would suggest keeping it like this as continued training always goes hand in hand with checkpointing in pytroch-lighting.

File: neuralprophet/forecaster.py [hunk 2]
```diff
@@ -2770,12 +2775,36 @@ def _train(
...
+            if self.config_train.epochs:
+                additional_epochs = self.config_train.epochs
+            else:
+                additional_epochs = previous_epoch
+            # Get the number of epochs already trained
+            new_total_epochs = previous_epoch + additional_epochs
+            self.config_train.epochs = new_total_epochs
+
+            # Reinitialize optimizer with loaded model parameters
+            optimizer = torch.optim.AdamW(self.model.parameters())
```
[the remark this answers is not in the document]
Comment (Constantin343) [reply neuralprophet/forecaster.py:2800]: I realized it is actually not necessary to reinitialize the optimizer like this and the regular flow that uses configure.py works as well.

**Comment by Constantin343** [comment]
> Furthermore, I think we should avoid hard-coding the optimizer and scheduler for this functionality. I suggest you make them optionally user-settable (with reasonable defaults). And if so, then it may make sense to do the same for the regular optimizer and scheduler in the same manner.

Totally agree. I tried again to make it work with the OneCycleLR scheduler, but it just doesn't seem to work with this scheduler. I will try to make some adaptations over the weekend to accept scheduler_type as a parameter, so the user can choose some scheduler types.
